src/backends/kimchi/prover.rs

```diff
@@ -233,3 +233,113 @@ impl VerifierIndex {
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        backends::{
+            kimchi::{KimchiVesta, VestaField},
+            Backend,
+        },
+        compiler::{compile, generate_witness, typecheck_next_file, Sources},
+        constants::Span,
+        inputs::parse_inputs,
+        type_checker::TypeChecker,
+        var::{ConstOrCell, Value, Var},
+        witness::WitnessEnv,
+    };
+
+    #[test]
+    fn test_public_output_constraint() -> miette::Result<()> {
+        let code = r#"fn main(pub public_input: Field, private_input: Field) -> Field {
+            let xx = private_input + public_input;
+            assert_eq(xx, 2);
+            let yy = xx + 6;
+            return yy;
+        }"#;
+
+        let mut sources = Sources::new();
+        let mut tast = TypeChecker::new();
+        let this_module = None;
+        let _node_id = typecheck_next_file(
+            &mut tast,
+            this_module,
+            &mut sources,
+            "inline_test_output.no".to_string(),
+            code.to_owned(),
+            0,
+        )
+        .unwrap();
+
+        let kimchi_vesta = KimchiVesta::new(false);
+        let compiled_circuit = compile(&sources, tast, kimchi_vesta)?;
+
+        let (prover_index, _) = compiled_circuit.compile_to_indexes().unwrap();
+
+        // parse inputs
+        let public_inputs = parse_inputs(r#"{"public_input": "1"}"#).unwrap();
+        let private_inputs = parse_inputs(r#"{"private_input": "1"}"#).unwrap();
+
+        let public_input_val: u64 = public_inputs
+            .0
+            .get("public_input")
+            .unwrap()
+            .as_str()
+            .unwrap()
+            .parse()
+            .unwrap();
+        let output_val = 8;
+
+        let generated_witness = generate_witness(
+            &prover_index.compiled_circuit,
+            &sources,
+            public_inputs,
+            private_inputs,
+        )?;
+
+        assert_eq!(
+            generated_witness.full_public_inputs,
+            vec![
+                VestaField::from(public_input_val),
+                VestaField::from(output_val)
+            ]
+        );
+
+        // modify the output value
+        let mut full_public_inputs = generated_witness.full_public_inputs.clone();
+        let mut witness = generated_witness.all_witness.to_kimchi_witness();
+        let gate_count = prover_index.compiled_circuit.circuit.backend.gates.len();
+        assert_eq!(gate_count, 6);
+        assert_eq!(witness[0][0], VestaField::from(public_input_val));
+        // this is the gate contains the output var
+        assert_eq!(witness[0][1], VestaField::from(output_val));
+        // this is the assert_eq gate which result var and the output var
+        assert_eq!(witness[0][gate_count - 1], VestaField::from(output_val));
+
+        // should pass the sanity check
+        prover_index
+            .index
+            .verify(&witness, &full_public_inputs)
+            .unwrap();
```

Comment:
can you also check that this returns an error as well if we modify the pub output then? (if what you say is correct about the permutation I guess this should work as well)

Reply:
it does this check here:
https://github.com/zksecurity/noname/blob/444ca6a2c8f19318f224ed64e93b315c96b938c4/src/backends/kimchi/prover.rs#L325-L341

perhaps we can also make it assert it returns `DisconnectedWires ` error (see the above code snippet) if the output is modified.